README.md
## bindata

This package converts any file into managable Go source code. Useful for
embedding binary data into a go program. The file data is optionally gzip
compressed before being converted to a raw byte slice.

It comes with a command line tool in the `go-bindata` sub directory.
This tool offers a set of command line options, used to customize the
output being generated.


### Installation

To install the library and command line program, use the following:

	go get -u github.com/shuLhan/go-bindata/...


### Usage

Conversion is done on one or more sets of files. They are all embedded in a new
Go source file, along with a table of contents and an `Asset` function,
which allows quick access to the asset, based on its name.

The simplest invocation generates a `bindata.go` file in the current
working directory. It includes all assets from the `data` directory.

	$ go-bindata data/

To include all input sub-directories recursively, use the elipsis postfix
as defined for Go import paths. Otherwise it will only consider assets in the
input directory itself.

	$ go-bindata data/...

To specify the name of the output file being generated, we use the following:

	$ go-bindata -o myfile.go data/

Multiple input directories can be specified if necessary.

	$ go-bindata dir1/... /path/to/dir2/... dir3


The following paragraphs detail some of the command line options which can be 
supplied to `go-bindata`. Refer to the `testdata/out` directory for various
output examples from the assets in `testdata/in`. Each example uses different
command line options.

To ignore files, pass in regexes using -ignore, for example:

    $ go-bindata -ignore=\\.gitignore data/...

### Accessing an asset

To access asset data, we use the `Asset(string) ([]byte, error)` function which
is included in the generated output.

	data, err := Asset("pub/style/foo.css")
	if err != nil {
		// Asset was not found.
	}

	// use asset data


### Debug vs Release builds

When invoking the program with the `-debug` flag, the generated code does
not actually include the asset data. Instead, it generates function stubs
which load the data from the original file on disk. The asset API remains
identical between debug and release builds, so your code will not have to
change.

This is useful during development when you expect the assets to change often.
The host application using these assets uses the same API in both cases and
will not have to care where the actual data comes from.

An example is a Go webserver with some embedded, static web content like
HTML, JS and CSS files. While developing it, you do not want to rebuild the
whole server and restart it every time you make a change to a bit of
javascript. You just want to build and launch the server once. Then just press
refresh in the browser to see those changes. Embedding the assets with the
`debug` flag allows you to do just that. When you are finished developing and
ready for deployment, just re-invoke `go-bindata` without the `-debug` flag.
It will now embed the latest version of the assets.


### Lower memory footprint

Using the `-nomemcopy` flag, will alter the way the output file is generated.
It will employ a hack that allows us to read the file data directly from
the compiled program's `.rodata` section. This ensures that when we call
call our generated function, we omit unnecessary memcopies.

The downside of this, is that it requires dependencies on the `reflect` and
`unsafe` packages. These may be restricted on platforms like AppEngine and
thus prevent you from using this mode.

Another disadvantage is that the byte slice we create, is strictly read-only.
For most use-cases this is not a problem, but if you ever try to alter the
returned byte slice, a runtime panic is thrown. Use this mode only on target
platforms where memory constraints are an issue.

The default behaviour is to use the old code generation method. This
prevents the two previously mentioned issues, but will employ at least one
extra memcopy and thus increase memory requirements.

For instance, consider the following two examples:

This would be the default mode, using an extra memcopy but gives a safe
implementation without dependencies on `reflect` and `unsafe`:

```go
func myfile() []byte {
    return []byte{0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a}
}
```

Here is the same functionality, but uses the `.rodata` hack.
The byte slice returned from this example can not be written to without
generating a runtime error.

```go
var _myfile = "\x89\x50\x4e\x47\x0d\x0a\x1a"

func myfile() []byte {
    var empty [0]byte
    sx := (*reflect.StringHeader)(unsafe.Pointer(&_myfile))
    b := empty[:]
    bx := (*reflect.SliceHeader)(unsafe.Pointer(&b))
    bx.Data = sx.Data
    bx.Len = len(_myfile)
    bx.Cap = bx.Len
    return b
}
```


### Optional compression

When the `-nocompress` flag is given, the supplied resource is *not* GZIP
compressed before being turned into Go code. The data should still be accessed
through a function call, so nothing changes in the usage of the generated file.

This feature is useful if you do not care for compression, or the supplied
resource is already compressed. Doing it again would not add any value and may
even increase the size of the data.

The default behaviour of the program is to use compression.


### Path prefix stripping

The keys used in the `_bindata` map are the same as the input file name passed
to `go-bindata`. This includes the path. In most cases, this is not desireable,
as it puts potentially sensitive information in your code base.  For this
purpose, the tool supplies another command line flag `-prefix`.  This accepts a
[regular expression](https://github.com/google/re2/wiki/Syntax) string, which
will be used to match a portion of the map keys and function names that should
be stripped out.

For example, running without the `-prefix` flag, we get:

	$ go-bindata /path/to/templates/

	_bindata["/path/to/templates/foo.html"] = path_to_templates_foo_html

Running with the `-prefix` flag, we get:

	$ go-bindata -prefix "/.*/some/" /a/path/to/some/templates/

	_bindata["templates/foo.html"] = templates_foo_html


### Build tags

With the optional `-tags` flag, you can specify any go build tags that
must be fulfilled for the output file to be included in a build. This
is useful when including binary data in multiple formats, where the desired
format is specified at build time with the appropriate tags.

The tags are appended to a `// +build` line in the beginning of the output file
and must follow the build tags syntax specified by the go tool.

### Related projects

[go-bindata-assetfs](https://github.com/elazarl/go-bindata-assetfs#readme) - 
implements `http.FileSystem` interface. Allows you to serve assets with `net/http`.



CHANGELOG
# v4.0.0 (2020-08-30)

This release fix the Go module path, following the best practice as described
in https://blog.golang.org/v2-go-modules.

The previous go.mod in the root directory was invalid, because we have version
v3.x.x tags long before we add the Go modules which make the "go get" command
can not fetch version greater than v3.4.0.

The root directory will be keep for backward compatibility, to be consumed by
users that still use GOPATH, and new user that want to use Go modules should
import "github.com/shuLhan/go-bindata/v4".


# v3.6.1 (2020-08-30)

Remove the go.mod from root repository.

After we add go.mod, the latest release version listed on pkg.go.dev
is v3.4.0 and version after that seems get rejected because Semantic
Import Versioning rules.


# v3.6.0 (2020-07-16)

##  New features

*  Add the ability to configure the asset prefix so it can be exported

By default, the asset function is not exported.  This changes add option to
set the function prefix with `assetprefix` option, so one can make them
exported.

##  Bug fixes

* Fix the default value for cfg.Split

* test: provide a modtime for each test case


# v3.5.2 (2020-05-03)

##  Bug fixes

*  Fix empty assets when running on directory without recursive (#38)

   Due to hasty fixes on #36 and #37, the bug #28 reappear again in issue #38.

##  Enhancements

*  all: refactor the unit tests

   Previous unit tests are really convoluted, its hard to read and its
   even hard to add new tests that reflect use cases in real worlds.
   This causes several issues reappear after fixing another issues.

   This changes rewrite the unit tests by replacing test using go-bindata
   binary with call to Translate() with custom Config depends on each unit
   test.

*  all: cleaning up API, un-export all internal types

*  all: use map to collect list of assets instead of slice


# v3.5.1 (2020-05-01)

##  Bug fixes

*  config: check for duplicate inputs (#37)
*  fsscanner: fix duplicate assets generated without recursive path (#36)


# v3.5.0 (2020-04-07)

##  Enhancements

*  Add verbose option (#21)

   The verbose option ("-verbose") print the file that being scanned.
   For file that is being added (either with include option or not) it will
   print the path prefixed with "+", and for file that is ignored it will
   print the path prefixed with "-".

   The last line will print the name of generated Go file.

##  Bug Fixes

*  Fix stack overflow when loading large data with compress option (#35)

*  Fix empty assets when running on directory without recursive (#28)

   This fix issue that result in empty generated Go when running `go-bindata`
   on directory without recursive ("...") directive.

##  Misc

*  test: fix test with Go version 1.14 or greater


# v3.4.0 (2018-05-19)

Linting task and dependencies is removed from make task. Linting should be
decoupled from any library or repository. It is a tool that not affected the
build, but help on detect bugs or analyze code.

## Enhancements

* @rsc: balance backquoted string concatenation
* Emad Elsaid: Remove duplicate words on README

## Bug Fixes

* @fgrosse: Fix using symbolic links


# v3.3.0 (2017-11-07)

This release is contain major internal refactoring on scanning the file system.

## Enhancements

* Check error and return it

  There are many place where error from file create, write, and close was not
  checked. With the help of tools, all error was checked and returned to the
  caller.

* Move generated content (template) to its own file

  Since we generate go code, the template become noise because its inlined
  with the generator itself, which make the code unreadable. This commit
  split and move the content of generated to its own file `template.go`.

* Refactoring the file scanner

  There are several problems in old file system scanner,

  * It's unreadable.  It's hard to read what the code do from the start
    until the end, which make it even harder to write new function/feature.

  * Each input file is converted to absolute path to handle symlink
    files and/or directory which cause additional process. The new scanner
    work with relative path only.

## Bug Fixes

* @joerocklin: Add the proper 'DO NOT EDIT' tag on file assets
* @komuw: update installation with the right location


# v3.2.0 (2017-11-07)

## New Features

* @ZaniaDeveloper 94068b1 -  Avoid having a big file (#17)

  This feature introduce `-split` option to config and command line. If its set
  the output file is splitted per asset file.

* @typeless 31e070ef - Add `-include` regex pattern to command line option

## Enhancements

* Rewrite makefile
* Remove and ignore testdata output, 'testdata/out', from repository
* Move "DO NOT EDIT" mark to the top, inline with "... generated by ..."

* release:
  * assetReleaseCommon: fix shadowed error variable
  * Rename underscores function name to snake-case

* [lint] Add comment to ByteWriter and StringWriter

* config:
  * Split validate() into validateInput() and validateOutput()
  * Error should not end with punctuation

* convert:
  * findFiles: never, ever use `defer` on recursive function
  * Fix comment on ByName type

* doc:
  * Fix misspellings
  * Fix package comment

* toc
  * Fix space alignment in generated asset names
  * writeGoMap: remove empty second value in range
  * funcOrNil: simplify if-else
  * child: use consistent receiver name

* go-bindata:
  * Use instance of package log to print to stdout and stderr
  * Refactoring parse arguments
  * Add comment to Set method
  * Split and move usage into function
  * Merge version to main
  * Split flag initialization
  * parsingInput:
    * add unit test
    * fix comment and simplify if-else

## Bug Fixes

* [test] convert: fix due to changes of prefix from string to regex


# v3.1.0 (2017-09-10)

## New Features

* @apvail 8f5f6db - Alter `-prefix` to Use Regular Expressions instead of just
  strings
* @akutz 9308d8f - Added MD5 Checksum Support

config.go
// Copyright 2018 The go-bindata Authors. All rights reserved.
// Use of this source code is governed by a CC0 1.0 Universal (CC0 1.0)
// Public Domain Dedication license that can be found in the LICENSE file.

package bindata

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
)

const (
	// DefPackageName define default package name.
	DefPackageName = "main"

	// DefOutputName define default generated file name.
	DefOutputName = "bindata.go"
)

// List of errors.
var (
	ErrNoInput       = errors.New("no input")
	ErrNoPackageName = errors.New("missing package name")
	ErrCWD           = errors.New("unable to determine current working directory")
)

// Config defines a set of options for the asset conversion.
type Config struct {
	// cwd contains current working directory.
	cwd string

	// Name of the package to use. Defaults to 'main'.
	Package string

	// Tags specify a set of optional build tags, which should be
	// included in the generated output. The tags are appended to a
	// `// +build` line in the beginning of the output file
	// and must follow the build tags syntax specified by the go tool.
	Tags string

	// Input defines the directory path, containing all asset files as
	// well as whether to recursively process assets in any sub directories.
	Input []InputConfig

	// Output defines the output file for the generated code.
	// If left empty, this defaults to 'bindata.go' in the current
	// working directory and the current directory in case of having true
	// to `Split` config.
	Output string

	// Prefix defines a regular expression which should used to strip
	// substrings from all file names when generating the keys in the table of
	// contents.  For example, running without the `-prefix` flag, we get:
	//
	// 	$ go-bindata /path/to/templates
	// 	go_bindata["/path/to/templates/foo.html"] = _path_to_templates_foo_html
	//
	// Running with the `-prefix` flag, we get:
	//
	//	$ go-bindata -prefix "/.*/some/" /a/path/to/some/templates/
	//	_bindata["templates/foo.html"] = templates_foo_html
	Prefix *regexp.Regexp

	// Ignores any filenames matching the regex pattern specified, e.g.
	// path/to/file.ext will ignore only that file, or \\.gitignore
	// will match any .gitignore file.
	//
	// This parameter can be provided multiple times.
	Ignore []*regexp.Regexp

	// Include contains list of regex to filter input files.
	Include []*regexp.Regexp

	// When nonzero, use this as mode for all files.
	Mode uint

	// When nonzero, use this as unix timestamp for all files.
	ModTime int64

	// When true, size, mode and modtime are not preserved from files
	NoMetadata bool

	// NoMemCopy will alter the way the output file is generated.
	//
	// It will employ a hack that allows us to read the file data directly from
	// the compiled program's `.rodata` section. This ensures that when we call
	// call our generated function, we omit unnecessary mem copies.
	//
	// The downside of this, is that it requires dependencies on the `reflect` and
	// `unsafe` packages. These may be restricted on platforms like AppEngine and
	// thus prevent you from using this mode.
	//
	// Another disadvantage is that the byte slice we create, is strictly read-only.
	// For most use-cases this is not a problem, but if you ever try to alter the
	// returned byte slice, a runtime panic is thrown. Use this mode only on target
	// platforms where memory constraints are an issue.
	//
	// The default behaviour is to use the old code generation method. This
	// prevents the two previously mentioned issues, but will employ at least one
	// extra memcopy and thus increase memory requirements.
	//
	// For instance, consider the following two examples:
	//
	// This would be the default mode, using an extra memcopy but gives a safe
	// implementation without dependencies on `reflect` and `unsafe`:
	//
	// 	func myfile() []byte {
	// 		return []byte{0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a}
	// 	}
	//
	// Here is the same functionality, but uses the `.rodata` hack.
	// The byte slice returned from this example can not be written to without
	// generating a runtime error.
	//
	// 	var _myfile = "\x89\x50\x4e\x47\x0d\x0a\x1a"
	//
	// 	func myfile() []byte {
	// 		var empty [0]byte
	// 		sx := (*reflect.StringHeader)(unsafe.Pointer(&_myfile))
	// 		b := empty[:]
	// 		bx := (*reflect.SliceHeader)(unsafe.Pointer(&b))
	// 		bx.Data = sx.Data
	// 		bx.Len = len(_myfile)
	// 		bx.Cap = bx.Len
	// 		return b
	// 	}
	NoMemCopy bool

	// NoCompress means the assets are /not/ GZIP compressed before being turned
	// into Go code. The generated function will automatically unzip
	// the file data when called. Defaults to false.
	NoCompress bool

	// Perform a debug build. This generates an asset file, which
	// loads the asset contents directly from disk at their original
	// location, instead of embedding the contents in the code.
	//
	// This is mostly useful if you anticipate that the assets are
	// going to change during your development cycle. You will always
	// want your code to access the latest version of the asset.
	// Only in release mode, will the assets actually be embedded
	// in the code. The default behaviour is Release mode.
	Debug bool

	// Perform a dev build, which is nearly identical to the debug option. The
	// only difference is that instead of absolute file paths in generated code,
	// it expects a variable, `rootDir`, to be set in the generated code's
	// package (the author needs to do this manually), which it then prepends to
	// an asset's name to construct the file path on disk.
	//
	// This is mainly so you can push the generated code file to a shared
	// repository.
	Dev bool

	// Split the output into several files. Every embedded file is bound into
	// a specific file, and a common file is also generated containing API and
	// other common parts.
	// If true, the output config is a directory and not a file.
	Split bool

	// MD5Checksum is a flag that, when set to true, indicates to calculate
	// MD5 checksums for files.
	MD5Checksum bool

	// Verbose flag to display verbose output.
	Verbose bool
}

//
// NewConfig returns a default configuration struct.
//
func NewConfig() *Config {
	c := new(Config)
	c.Package = DefPackageName
	c.Output = DefOutputName
	c.Ignore = make([]*regexp.Regexp, 0)
	c.Include = make([]*regexp.Regexp, 0)
	return c
}

func (c *Config) validateInput() (err error) {
	uniqPaths := make(map[string]struct{}, len(c.Input))
	newInputs := make([]InputConfig, 0, len(c.Input))

	for _, input := range c.Input {
		input.Path = filepath.Clean(input.Path)
		_, ok := uniqPaths[input.Path]
		if ok {
			continue
		}
		_, err = os.Lstat(input.Path)
		if err != nil {
			return fmt.Errorf("failed to stat input path '%s': %v",
				input.Path, err)
		}
		uniqPaths[input.Path] = struct{}{}
		newInputs = append(newInputs, input)
	}
	if len(newInputs) == 0 {
		return ErrNoInput
	}
	c.Input = newInputs
	return nil
}

//
// validateOutput will check if output is valid.
//
// (1) If output is empty, set the output directory to,
// (1.1) current working directory if `split` option is used, or
// (1.2) current working directory with default output file output name.
// (2) If output is not empty, check the directory and file write status.
//
func (c *Config) validateOutput() (err error) {
	// (1)
	if len(c.Output) == 0 {
		if c.Split {
			// (1.1)
			c.Output = c.cwd
		} else {
			// (1.2)
			c.Output = filepath.Join(c.cwd, DefOutputName)
		}

		return nil
	}

	// (2)
	dir, file := filepath.Split(c.Output)

	if dir != "" {
		err = os.MkdirAll(dir, 0700)
		if err != nil {
			return fmt.Errorf("create output directory: %v", err)
		}
	}

	if len(file) == 0 {
		if !c.Split {
			c.Output = filepath.Join(dir, DefOutputName)
		}
	}

	if c.Split {
		return nil
	}

	var fout *os.File

	fout, err = os.Create(c.Output)
	if err != nil {
		return err
	}

	return fout.Close()
}

//
// validate ensures the config has sane values.
// Part of which means checking if certain file/directory paths exist.
//
func (c *Config) validate() (err error) {
	if len(c.cwd) == 0 {
		c.cwd, err = os.Getwd()
		if err != nil {
			return ErrCWD
		}
	}

	if len(c.Package) == 0 {
		return ErrNoPackageName
	}

	err = c.validateInput()
	if err != nil {
		return
	}

	err = c.validateOutput()
	if err != nil {
		return
	}

	return
}
